filters/valve.cpp

```diff
@@ -0,0 +1,207 @@
+  if (decodedValue.isSome() && decodedValue.get().empty()) {
+    return Error("Empty string for '" + key + "'.");
+  }
+
+  return decodedValue.get();
+}
+
+
+class ValveFilterEndpointProcess : public Process<ValveFilterEndpointProcess> {
+ public:
+  ValveFilterEndpointProcess(ValveType _valveType, bool _opened)
+    : ProcessBase(getProcessBaseName(_valveType)),
+      valveType(_valveType),
+      opened(_opened),
+      limiter(2, Seconds(1)) {}  // 2 permits per second.
+
+  virtual ~ValveFilterEndpointProcess() {}
+
+  void safeOpen(bool open) {
+    synchronized(control_mutex) {
+      // NOTE: In future we may want to trigger some actions here.
+      this->opened = open;
+    }
+  }
+
+  Future<bool> isOpened() {
+    return this->opened;
+  }
+
+  const lambda::function<Future<bool>()> getIsOpenedFunction() {
+    return defer(self(), &Self::isOpened);
+  }
+
+ protected:
+  virtual void initialize() {
+    switch (valveType) {
+      case RESOURCE_ESTIMATOR_VALVE:
+        route(VALVE_ROUTE,
+              ESTIMATOR_VALVE_ENDPOINT_HELP(),
+              &ValveFilterEndpointProcess::valve);
+        LOG(INFO) << "Serenity Resource Estimator endpoint initialized "
+                  << "on /" << RESOURCE_ESTIMATOR_PROCESS_BASE  << VALVE_ROUTE;
+        break;
+      case QOS_CONTROLLER_VALVE:
+        route(VALVE_ROUTE,
+              CONTROLLER_VALVE_ENDPOINT_HELP(),
+              &ValveFilterEndpointProcess::valve);
+        LOG(INFO) << "Serenity QoS Controller endpoint initialized "
+                  << "on /" << QOS_CONTROLLER_PROCESS_BASE  << VALVE_ROUTE;
+        break;
+    }
+  }
+
+ private:
+  Future<http::Response> valve(const http::Request& request) {
+    return limiter.acquire()
+      .then(defer(self(), &Self::_valve, request));
+  }
+
+  Future<http::Response> _valve(const http::Request& request) {
+    Try<hashmap<string, string>> decode =
+        process::http::query::decode(request.body);
+    if (decode.isError()) {
+      return http::BadRequest(
+          "Unable to decode query string: " + decode.error());
+    }
+    hashmap<string, string> values = decode.get();
+
+    // Get params.
+    string enabled_param;
+    Option<string> pipeline_enable = request.query.get(PIPELINE_ENABLE_KEY);
+    if (pipeline_enable.isSome()) {
+      enabled_param = pipeline_enable.get();
+    } else {
+      Try<string> pipeline_enable_form =
+          getFormValue(PIPELINE_ENABLE_KEY, values);
+
+      if (pipeline_enable_form.isError()) {
+        return http::BadRequest(pipeline_enable_form.error());
+      }
+      enabled_param = pipeline_enable_form.get();
+    }
+
+    bool pipeline_enable_decision = false;
+    if (!enabled_param.compare("true")) {
+      pipeline_enable_decision = true;
+    }
+
+    this->safeOpen(pipeline_enable_decision);
+
+    JSON::Object response;
+    response.values[PIPELINE_ENABLE_KEY] = enabled_param;
+
+    return http::OK(response);
+  }
+
+  bool opened;
+  Mutex control_mutex;
```

Comment:
Instead of mutex over bool, you should go for std::atomic<bool>.
No mutexes, no deadlocks, no problems.